persistence-cassandra-4.0/src/main/java/io/stargate/db/cassandra/impl/Cassandra40Persistence.java

```diff
@@ -500,12 +509,22 @@ protected void loginInternally(io.stargate.db.AuthenticatedUser user) {
 
     @Override
     public CompletableFuture<Result.Prepared> prepare(String query, Parameters parameters) {
-      return executeRequestOnExecutor(
-          parameters,
-          // The queryStartNanoTime is not used by prepared message, so it doesn't really matter
-          // that it's only computed now.
-          System.nanoTime(),
-          () -> new PrepareMessage(query, parameters.defaultKeyspace().orElse(null)));
+      // The behavior of prepared statements was changed as part of CASSANDRA-15252
+      // (and CASSANDRA-17248) which can cause multiple prepares of the same query to evict each
+      // other from the prepared cache. A few Stargate APIs eagerly prepare queries which
+      // increases this chance and to avoid this we serialize prepares for similar queries.
+      Lock lock = PREPARE_LOCKS.get(parameters.defaultKeyspace().map(k -> query + k).orElse(query));
```

Comment:
So this calculates hash for striped locks from query string, qualified by default namespace, if any?

Although purely for bucketing of potentially conflicting queries into same bucket, wouldn't it be enough to only consider `query` and ignore default keyspace? Should still use locking where it needs to be, unlikely to cause less congestion, and avoids String concatenation? So:

```
Lock lock = PREPARE_LOCKS.get(query)
```

Not a big deal but just wanted to see if my understanding is correct.

Reply:
Yeah, i think that make sense.  It would only become contentious when the same query strings are prepared in several different keyspaces. Which is probably not likely? I'm not sure.  I'm fine with just choosing the bucket based on `query`.  

It might be worth avoiding the concatenate for the common case to avoid some contention on the less likely case,

Reply:
I have no data to back this up so I am just speculating.

I am ok with using concatenation too, not a big deal, so it's up to you.